machinelearning/tensorflow.md

```diff
@@ -0,0 +1,405 @@
+# ML inference on Graviton CPUs with TensorFlow
+
+**Introduction**
+
+TensorFlow is an open-source software library for machine learning and artificial intelligence. It can be used across training and inference of deep neural networks. This document covers how to use Tensorflow based machine learning inference on Graviton CPUs, what runtime configurations are important and how to debug any performance issues. The document also covers instructions for source builds and how to enable some of the downstream features.
+
+# How to use Tensorflow on Graviton CPUs
+To start with how to use Tensorflow inference on Graviton CPUs, there are multiple levels of software package abstractions available: Python wheel (quick option), Docker container (comes with wheel, additional packages and benchmarks) and DLAMI (Deep Learning Amazon Machine Image that comes with preinstalled packages and tools). Let's see how one can use each of them to get started.
+
+**Using Python wheel**
+Tensorflow official supports optimized onednn+acl backend for Graviton CPUs.
+```
+pip install tensorflow-cpu-aws
+```
+
+**Using Docker hub container**
+
+Launch a Graviton instance with an Ubuntu 20.04 AMI and install docker using the below instructions.
+```
+#install and enable docker
+sudo apt update
+sudo apt install -y docker.io
+sudo systemctl enable docker && sudo systemctl start docker && sudo usermod -a -G docker ubuntu
+sudo chmod 666 /var/run/docker.sock
+
+# pull the tensorflow docker container with onednn-acl optimizations enabled
+docker pull armswdev/tensorflow-arm-neoverse
+
+# launch the docker image
+docker run -it --rm -v /home/ubuntu/:/hostfs armswdev/tensorflow-arm-neoverse
+```
+
+**Using AWS DLAMI**
+
+To launch a Graviton instance with a DLAMI via EC2 console:
+
+1. Click on "Community AMIs"
+2. Filter AMIs by clicking Ubuntu and Arm (64-bit) check boxes
+3. Search for "Deep Learning" in search field to list the current Ubuntu based DLAMIs"
+4. Select the AMI: "Deep Learning AMI Graviton Tensorflow <version>(Ubuntu 20.04) <yyyy/mm/dd>"
+
+Here is the awscli snippet to find the latest AMI:
+```
+aws ec2 describe-images --region us-west-2 --filters  Name=architecture,Values=arm64 Name=name,Values="*Deep Learning AMI*" Name=owner-alias,Values=amazon  --query 'Images[] | sort_by(@, &CreationDate)[-1] | ImageId'
+```
+
+Here is the awscli snippet to launch the same:
+```
+aws ec2 run-instances --region us-west-2 --image-id <ami-id from the above command>  --instance-type c7g.4xlarge --count 1  --key-name <key name> --subnet-id <subnet_id> --block-device-mappings 'DeviceName=/dev/sda1,Ebs={VolumeSize=256}'
```

Comment:
Can we make this a one-liner instead of two steps?

Reply:
ok